Render

BenoitRoux0 opened the pull request:
Merge render onto master

gd-harco committed:
Add drawing functionality to the project

Introduced new drawing functions that will aid in rendering visuals for this project. This includes creating a line, pixel placement, drawing a line on an image and handling different slopes for line drawing. Also made modifications to some existing implementations to make them compatible. This effort will provide a foundation for constructing visuals in later stages of the project.

BenoitRoux0 committed:
Add raycasting functionality for game rendering

This new file titled raycast.c was introduced to perform raycasting as part of the rendering process for the game. Specifically, it includes functions for casting vertical and horizontal rays, a ray-collision function, and a method for performing "Digital Differential Analysis" (dda). These implementations together create a 2D projection of a 3D environment within the game. The associated headers, render.h and cub.h, were updated for compatibility, and raycasting was incorporated into the primary rendering process within the render.c file.

BenoitRoux0 committed:
Add defines.h to project structure and to Makefile

A new header file defines.h was created to organize and centralize constants used across the project. This file includes definitions for window dimensions, cell size, player speed, and other parameters. These constants were previously hard-coded in multiple places making the project hard to maintain or update. The Makefile was also updated to include the new defines.h file in the compilation process.

BenoitRoux0 committed:
Add new image handling functionalities and separate `srcs`

- Adds img.h header file with new color, image and tileset structures
- New color handling options, like alpha channel, conversion between `t_color` and `uint32_t` color types
- Enumerate functions for creating, modifying and destroying `t_uint_img `and `t_mlx_img` types
- Tileset structures definition along with initialization and deletion functions
- Flip functions for horizontal and vertical image orientations
- New source files (C files) for img folder with these functionalities implemented
- `Make` now includes `img` directory sources with new `sources.mk `

These changes were necessary for better handling of colors and images in the code. The improved granular handling of these aspects, as well as the code separation into more source files, promotes better readability and compliance with solid principles.

BenoitRoux0 committed:
Add player control movement and updates

The commit introduces several new files and methods that handle player control and movements in the game. `input_player.c` and `update_player.c` have been created to manage keypress events and update player's position respectively. Another file `init_player.c` is created to initialize the player's properties. Changes are also made in `hooks.c` to accommodate the keypress and key release events. The `sources.mk` file is updated to include these files into the main build system.

BenoitRoux0 committed:
Add player control movement and updates

The commit introduces several new files and methods that handle player control and movements in the game. `input_player.c` and `update_player.c` have been created to manage keypress events and update player's position respectively. Another file `init_player.c` is created to initialize the player's properties. Changes are also made in `hooks.c` to accommodate the keypress and key release events. The `sources.mk` file is updated to include these files into the main build system.

BenoitRoux0 committed:
solve merge

BenoitRoux0 committed:
In the 'input_player.c' file, the conditions to alter the player's movements have been revised to correct the movement logic. This change was brought about due to incorrect movement directions - the keys mapped for each direction were leading to the opposite movements.

In the 'update_player.c' file, refined the player's angle and position update logic by adding more conditions and calculations, involving the player's current angle and speed. Also, debug logs were commented out to clean up the console output when the game is running. This change was done to make the character's movements more precise and natural, depending on the speed and orientation.

BenoitRoux0 committed:
Revise raycasting calculations and movement logic

In the 'raycast.c' file, functions for vertical and horizontal collision checks have been separated to rectify the raycasting routine for more accurate wall detection. The changes were mainly brought to correct errors in the wall collision checks which were causing aberrations in rendering the layout of the map.

Further, the current_angle variable, which calculates the direction of the player's sight based on his current orientation and Field of View, has been modulated correctly to maintain within the range of 0 to 360 degrees.

The raycasting routine has also been redefined by switching from global rays to individual horizontal and vertical rays, enabling accurate rendering of vertical and horizontal walls. This change was made as the previous implementation was providing incorrect distances due to incorrect angle calculations, leading to distorted rendering of the walls.

Also, verbose debugging lines have been commented out for cleaner console outputs.

BenoitRoux0 committed:
merge render

BenoitRoux0 committed:
Add player controls and modify relevant structures

Created `player.h` file to define player structure and functions to handle player controls. Modified `types.h` to adjust struct definitions including the addition of 'player' and 'rays', and enumeration 'e_dir' to manage the player's direction. Implementations for key press and release events were added to 'hooks.h'.

Adjusted certain parameters in 'defines.h' in order to align with the introduced player functionality. Furthermore, added necessary changes in 'img.h' and 'cub.h' to cater adjustments in the struct definition and function parameters.

These edits was necessary in order to organize code, manage the player's movement and actions easily and improve the encapsulation. For instance, handling all game related structures in 'types.h' is more maintainable that scattering each structure into separate files. Tweaking the 'defines.h' allows fine-tuning of game's parameters easier.

BenoitRoux0 committed:
Refactor variable and structure names for clarity

Changes in the code mainly involve renaming variables and structures for the sake of clarity and better understanding. The variable `bpp` was renamed to `bits_per_pixel` throughout the code to make its purpose clearer.

Similarly, the structure `t_mlx_img` was renamed to `t_master_img`, defining the master image more explicitly. This change is reflected throughout the corresponding functions parameters and calls.

This refactoring was necessary to improve readability and maintainability of the code. With more descriptive names, other developers

BenoitRoux0 committed:
Correct the variable name and improve raycasting logic

There is a correction in `draw.c` to modify a variable name (`img->line_length` to `img->line_len`). Additionally, there are various changes to the `raycast.c` file.

The changes introduce a more robust ray casting logic. The adjustments include type changes to function parameters, math computation adjustments, code clean-ups, and a new function `select_ray()` to handle ray selection logic between horizontal and vertical intersections.

These changes aim to improve the drawing and rendering capabilities and ensure the accuracy of the implementation. These will enhance the visual rendering and interaction within the game environment.

BenoitRoux0 committed:
Correct the variable name and improve raycasting logic

There is a correction in `draw.c` to modify a variable name (`img->line_length` to `img->line_len`). Additionally, there are various changes to the `raycast.c` file.

The changes introduce a more robust ray casting logic. The adjustments include type changes to function parameters, math computation adjustments, code clean-ups, and a new function `select_ray()` to handle ray selection logic between horizontal and vertical intersections.

These changes aim to improve the drawing and rendering capabilities and ensure the accuracy of the implementation. These will enhance the visual rendering and interaction within the game environment.

BenoitRoux0 committed:
Merge branch 'render' of github.com:BenoitRoux0/cub3D into render

BenoitRoux0 committed:
Refactor rendering and raycasting in main.c and raycast.c

The modifications in `main.c` include:
- Adding `img.h` to the import statements.
- Adding static functions `draw_map`, `draw_player`, and `loop_hook`.
- Updating the `main` function to use the new functions and initialize the necessary variables.

The modifications in `raycast.c` include:
- Changing the signature of the `dda` function to accept a `t_ray` struct as the starting point.
- Making adjustments to the `cast_vert` and `cast_horiz` functions to use the new `dda` function.
- Removing a commented out print statement in `ray_collide_vert`.
- Adding a commented out return statement in `select_ray`.

Note: The changes related to imports are ignored as per the instructions.

BenoitRoux0 committed:
Update drawing parameters and enhance raycasting

Improved rendering efficiency and visual representation by updating the 'draw.c' and 'raycast.c' files. In 'draw.c', renamed 'img->line_length' to 'img->line_len' for consistency and readability. Introduced better raycasting logic in 'raycast.c' with an adjustment to function parameters, mathematical computations, and code cleanliness. Added 'select_ray()' function for more efficient ray selection between horizontal and vertical intersections. Aimed to improve the game's drawing and rendering functionality, resulting in a more accurate and visually appealing gaming environment.

TO FIX: Wall height formula and remove fish eye effect

BenoitRoux0 committed:
merge render

BenoitRoux0 committed:
Update XPM image opening functions

Key changes are made to two files: 'ft_xpm_to_img.c' and 'get_textures.c'. Now, instead of returning a pointer to an image, the 'ft_xpm_to_img' function returns the image itself. In case of any error, it now returns an instance of 't_uint_img' with NULL content value, instead of pointer from the heap. This is intended to improve memory handling by minimizing potential leaks. Similarly, 'get_textures.c' now uses 'ft_xpm_to_img' function to directly acquire the image avoiding complexities linked with pointers. Also the Makefile is edited to link source code files with libft header files for better modularity and reusability.

BenoitRoux0 committed:
Upgrade map and player rendering with raycasting

This commit implements fundamental rudimentary rendering for both the map and the player in the 3D game engine. The additions in hooks.c establish a simple 2D view of the map and player using the draw_map and draw_player functions in the game loop. The raycast.c file is updated to correct the handling of ray intersections beyond the map borders and to create precision in detecting wall hits. Consequently, render.c is adjusted to draw each slice of wall hit by the rays to construct a 3D projection for the player. This is a pivotal upgrade from the existing flat map view and sets the foundation for further enhancements of the game engine's graphical capabilities.

gd-harco commented on srcs/player/input_player.c:
modifier les keycodes, passer en QWERTY

gd-harco commented on incs/types.h:
```diff
@@ -38,23 +38,61 @@ enum
+{
+	double	inter[2];
+	double	dist;
+	int		hit;
```
mettre un Boolean pour une meilleur lisibilte non ?

gd-harco commented on Makefile:
```diff
@@ -15,7 +15,11 @@ NAME =	cub3D
 INCS =	incs/cub.h
 
-include	srcs/render/sources.mk srcs/parsing/sources.mk
+include	srcs/render/sources.mk
+include srcs/img/sources.mk
+include srcs/player/sources.mk
+include	srcs/drawing/drawing.mk
+include srcs/parsing/sources.mk
 
```
pourquoi pa s plutot
includes ..... \
         ...... \
         ...... \

BenoitRoux0 committed:
Revise hit data type in Ray struct and improve player controls

The data type for the "hit" variable in the Ray struct has been changed from int to bool for clearer and more efficient checks of hit statuses. A field for image structure has also been added in the Data struct for facilitating the manipulation and storage of image data.

In addition, player controls have been updated to standard WASD layout from the previous ZQSD layout for better game experience and conformation to popular gaming standards.

BenoitRoux0 committed:
Update Makefile include statements formatting

Changed the formatting of include statements in the Makefile for cleaner code and easier readability. The includes sequences are now arranged vertically and separated by a backslash, improving the overall organization and readability of the code.
